File: examples/python/add_checklist_item.py
#!/usr/bin/env python3
"""
Notorious API - Add a checklist item example (Python stdlib only).

A checklist is a single block whose content holds ALL of its items
(`{"items": [{"id", "markdown", "checked"}, ...]}`) - items are not
separate blocks. PATCH /api/v1/blocks/:id shallow-merges the given
`content` into the existing one, so appending an item means sending the
full items array back with the new one added.

Usage:
    python add_checklist_item.py "Buy milk"
"""

import json
import sys
import urllib.error
import urllib.request
import uuid

API_URL = "http://localhost:4000"
API_TOKEN = "ntr_your_api_token_here"
OBJECT_ID = "your-object-id"
CHECKLIST_BLOCK_ID = "your-checklist-block-id"


def request_json(method: str, path: str, body: dict | None = None):
    data = json.dumps(body).encode() if body is not None else None
    request = urllib.request.Request(
        f"{API_URL}{path}",
        data=data,
        method=method,
        headers={
            "Authorization": f"Bearer {API_TOKEN}",
            "Content-Type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(request) as response:
            return json.loads(response.read())
    except urllib.error.HTTPError as error:
        body_text = json.loads(error.read())
        raise SystemExit(f"Request failed ({error.code}): {body_text.get('message')}")


def main() -> None:
    if len(sys.argv) != 2:
        raise SystemExit(f"Usage: python {sys.argv[0]} <item text>")
    item_text = sys.argv[1]

    blocks = request_json("GET", f"/api/v1/objects/{OBJECT_ID}/blocks")
    checklist = next((b for b in blocks if b["id"] == CHECKLIST_BLOCK_ID), None)
    if checklist is None or checklist["type"] != "checklist":
        raise SystemExit("CHECKLIST_BLOCK_ID does not point to a checklist block in OBJECT_ID.")

    new_item = {"id": str(uuid.uuid4()), "markdown": item_text, "checked": False}
    items = checklist["content"].get("items", []) + [new_item]

    request_json(
        "PATCH",
        f"/api/v1/blocks/{CHECKLIST_BLOCK_ID}",
        {"content": {"items": items}},
    )
    print(f"Added item: {item_text}")


if __name__ == "__main__":
    main()

File: examples/php/add_checklist_item.php
<?php
/**
 * Notorious API - Add a checklist item example (PHP curl extension only).
 *
 * A checklist is a single block whose content holds ALL of its items
 * (`{"items": [{"id", "markdown", "checked"}, ...]}`) - items are not
 * separate blocks. PATCH /api/v1/blocks/:id shallow-merges the given
 * `content` into the existing one, so appending an item means sending the
 * full items array back with the new one added.
 *
 * Usage:
 *   php add_checklist_item.php "Buy milk"
 */

const API_URL = 'http://localhost:4000';
const API_TOKEN = 'ntr_your_api_token_here';
const OBJECT_ID = 'your-object-id';
const CHECKLIST_BLOCK_ID = 'your-checklist-block-id';

function requestJson(string $method, string $path, ?array $body = null): mixed
{
    $ch = curl_init(API_URL . $path);
    $headers = ['Authorization: Bearer ' . API_TOKEN, 'Content-Type: application/json'];
    $options = [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_CUSTOMREQUEST => $method,
        CURLOPT_HTTPHEADER => $headers,
    ];
    if ($body !== null) {
        $options[CURLOPT_POSTFIELDS] = json_encode($body);
    }
    curl_setopt_array($ch, $options);
    $responseBody = curl_exec($ch);
    $status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    $data = json_decode($responseBody, true);
    if ($status >= 400) {
        fwrite(STDERR, "Request failed ({$status}): {$data['message']}\n");
        exit(1);
    }
    return $data;
}

function uuidV4(): string
{
    $bytes = random_bytes(16);
    $bytes[6] = chr((ord($bytes[6]) & 0x0f) | 0x40);
    $bytes[8] = chr((ord($bytes[8]) & 0x3f) | 0x80);
    $hex = bin2hex($bytes);
    return sprintf(
        '%s-%s-%s-%s-%s',
        substr($hex, 0, 8),
        substr($hex, 8, 4),
        substr($hex, 12, 4),
        substr($hex, 16, 4),
        substr($hex, 20, 12)
    );
}

function main(array $argv): void
{
    if (count($argv) !== 2) {
        fwrite(STDERR, "Usage: php {$argv[0]} <item text>\n");
        exit(1);
    }
    $itemText = $argv[1];

    $blocks = requestJson('GET', '/api/v1/objects/' . OBJECT_ID . '/blocks');

    $checklist = null;
    foreach ($blocks as $block) {
        if ($block['id'] === CHECKLIST_BLOCK_ID) {
            $checklist = $block;
            break;
        }
    }
    if ($checklist === null || $checklist['type'] !== 'checklist') {
        fwrite(STDERR, "CHECKLIST_BLOCK_ID does not point to a checklist block in OBJECT_ID.\n");
        exit(1);
    }

    $items = $checklist['content']['items'] ?? [];
    $items[] = ['id' => uuidV4(), 'markdown' => $itemText, 'checked' => false];

    requestJson('PATCH', '/api/v1/blocks/' . CHECKLIST_BLOCK_ID, [
        'content' => ['items' => $items],
    ]);

    echo "Added item: {$itemText}\n";
}

main($argv);
